checkup: Introduce target Namespace

orelmisan opened the pull request:
Currently, checkups are executed in ephemeral Namespace objects.

This change enables checkup execution in a target Namespace - the Namespace in which the user-supplied ConfigMap object is at.

Create and delete an ephemeral namespace only if the user-supplied ConfigMap object is at the "kiagnose" namespace.
Else, setup all objects in the target Namespace.

Signed-off-by: Orel Misan <omisan@redhat.com>

This PR implements the design document published in PR #126.

Followup PRs will:
1. Split checkup_test.go to test cases regarding ephemeral and target namespace scenarios.
2. Add failure test cases for the target namespace feature.


~~This PR depends on PR #140.~~

EdDev commented on kiagnose/internal/checkup/checkup.go:
```diff
@@ -292,6 +309,10 @@ func (c *Checkup) Teardown() error {
 
+func (c *Checkup) teardownWithTargetNamespace() error {
+	return nil
+}
```
I am not clear why we need this.

orelmisan replied:
So Kiagnose will not delete the target namespace.
In the next commits it holds the teardown logic, which is different from the current teardown logic.

EdDev replied:
I just saw it is used in the next commit to add more content to it.

orelmisan replied:
Moved it to the first commit which is doing a teardown action.

EdDev commented on kiagnose/internal/checkup/checkup_test.go:
```diff
@@ -179,6 +245,38 @@ func TestCheckTeardownShouldSucceed(t *testing.T) {
+
+	_, err := testClient.CoreV1().Namespaces().Create(context.Background(), targetNs, metav1.CreateOptions{})
+	assert.NoError(t, err)
+	assertNamespaceExists(t, testClient.Clientset, testTargetNs)
```
I do not think you need to validate that the `Create` succeeds beyond asserting its returned error.

orelmisan replied:
Ok.

orelmisan replied:
Done.

EdDev commented on kiagnose/internal/checkup/checkup_test.go:
```diff
@@ -179,6 +245,38 @@ func TestCheckTeardownShouldSucceed(t *testing.T) {
+	assert.NoError(t, testCheckup.Setup())
+
+	expectedEphemeralNamespaceName := nameGen.Name(checkup.EphemeralNamespacePrefix)
+	assertNamespaceDoesntExists(t, testClient.Clientset, expectedEphemeralNamespaceName)
```
How is the ephemeral namespace relevant here? I was under the impression this test checks the explicit target-ns.

orelmisan replied:
I wanted to check that the ephemeral namespace is not created by mistake.

EdDev reviewed (changes requested):
Review of the 1st commit.

orelmisan commented on kiagnose/internal/checkup/checkup.go:
```diff
@@ -90,7 +97,7 @@ func New(c kubernetes.Interface, name string, checkupConfig *config.Config, name
 		client:              c,
 		teardownTimeout:     defaultTeardownTimeout,
-		namespace:           NewNamespace(nsName),
+		namespace:           ephemeralNamespace,
```
Agreed.

orelmisan replied:
Done.

orelmisan commented on kiagnose/internal/checkup/checkup.go:
```diff
@@ -218,14 +225,16 @@ func (c *Checkup) Setup() error {
-	defer func() {
-		if err != nil {
-			_ = namespace.DeleteAndWait(c.client, c.namespace, c.teardownTimeout)
+	if c.namespace != nil {
```
Thank you for the suggestion.
For now I rather keep thinks simpler, but we can do it in a follow-up PR if you think it has benefits.

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/3373c7eb0a35752234bc9cd51e928550d63aaa6f..afddb3640770147736b8c8afa41509babd825902): Renamed `Checkup.namespace` to `Checkup.ephemeralNamespace`.
Removed two assertions that a namespace exists after its creation in unit tests.

orelmisan commented:
Rebase.

RamLavi reviewed (changes requested):
added some comments,
But more importantly, did we explore the option to use OwnerReference?

orelmisan commented:
> added some comments, But more importantly, did we explore the option to use OwnerReference?

Most of this PR is handling the tear down process when a checkup executes in a target namespace.
As the commit progress, it can be seen that a certain order of tear down should be in place (in order to strip the privileges from the checkup):
1. Delete the checkup Job and its underlying Pod.
2. Delete the ClusterRoleBindings.
3. Delete the RoleBinding to the ConfigMapWriter Role.
4. Delete the ServiceAccount.
5. Delete the ConfigMapWriter Role.

AFAIU, when using OwnerReference we don't have control over the order in which objects are deleted.

orelmisan commented on kiagnose/internal/checkup/checkup.go:
```diff
@@ -333,6 +333,10 @@ func (c *Checkup) teardownWithTargetNamespace() error {
 		errs = append(errs, err)
 	}
 
+	if err := configmap.Delete(c.client, c.resultConfigMap.Namespace, c.resultConfigMap.Name); err != nil {
```
Prior to reaching this place in the code, the results were fetched, the Job was deleted and all of the RBAC related objects were removed.

orelmisan replied:
Done.

orelmisan commented:
Added documentation regarding the feature in the main README file, and in the checkups' README files.

EdDev commented on kiagnose/internal/checkup/job/job.go:
```diff
@@ -76,6 +76,43 @@ func WaitForJobToFinish(client kubernetes.Interface, job *batchv1.Job, timeout t
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	deletePolicy := metav1.DeletePropagationForeground
```
I must admit, I am not sure if this is considered safe (or safer from the default background option).
In any case, I would suggest to say a few words about this in the commit, explaining your reasoning you choose it.

orelmisan replied:
Added reasoning to the commit message.
Can you please elaborate why do you think it is not considered safe?

EdDev replied:
It may be unsafe because it requires the dependent objects to be deleted first, which may actually fail and therefore stuck this operation, or even cause it to fail.

If something else blocks the deletion of the dependent, it is a deadlock.

Said all that, I am not sure what is the best approach in general. It may be good enough here.
The explanation you added to the commit seems convincing.
For the long run, some thought about how to delete things and the risks of each approach may need exploration.

EdDev reviewed (commented):
Looks very nice!

orelmisan commented:
> > vm latency checkup, e2e: Run checkup in a target namespace
> 
> IIUC this means that we no longer check the ephemeral NS usage, right? If this is the case, please mention it in the commit message.

Done.

orelmisan commented on kiagnose/internal/checkup/checkup.go:
```diff
@@ -310,6 +310,17 @@ func (c *Checkup) teardownWithEphemeralNamespace() error {
 }
 
 func (c *Checkup) teardownWithTargetNamespace() error {
+	const errPrefix = "teardown"
```
Done.

orelmisan commented on kiagnose/internal/checkup/checkup.go:
```diff
@@ -310,6 +310,17 @@ func (c *Checkup) teardownWithEphemeralNamespace() error {
+	var errs []error
+
+	if err := job.DeleteAndWait(c.client, c.job, c.teardownTimeout); err != nil {
+		errs = append(errs, err)
```
Moved the usage of `var errs []error` to the first commit that needs more than one error.

orelmisan commented on kiagnose/internal/checkup/checkup.go:
```diff
@@ -274,14 +283,22 @@ func (c *Checkup) SetTeardownTimeout(duration time.Duration) {
 }
 
 func (c *Checkup) Teardown() error {
+	if c.ephemeralNamespace != nil {
```
Done.

orelmisan commented:
Changes [1](https://github.com/kiagnose/kiagnose/compare/7bec0bd191fc1a3407da46314b787ed8ac33a946..f84b9d171cd87ca24738467ceb4bb60d0812aa06) and [2](https://github.com/kiagnose/kiagnose/compare/f84b9d171cd87ca24738467ceb4bb60d0812aa06..69447cf693152844c5b05462725c4423d7492a31): Answered review comments by @RamLavi and @EdDev.

RamLavi commented:
> > added some comments, But more importantly, did we explore the option to use OwnerReference?
> 
> Most of this PR is handling the tear down process when a checkup executes in a target namespace. As the commit progress, it can be seen that a certain order of tear down should be in place (in order to strip the privileges from the checkup):
> 
> 1. Delete the checkup Job and its underlying Pod.
> 2. Delete the ClusterRoleBindings.
> 3. Delete the RoleBinding to the ConfigMapWriter Role.
> 4. Delete the ServiceAccount.
> 5. Delete the ConfigMapWriter Role.
> 6. Delete the results ConfigMap.
> 
> AFAIU, when using OwnerReference we don't have control over the order in which objects are deleted.

I would try nonetheless just to see, it would make the teardown responsibilities out of our hands. 
Even if it does needs ordering, perhaps if we ever let go of the RBAC roles responsibilities then we won't need to worry about that.

RamLavi commented:
> I would try nonetheless just to see, it would make the teardown responsibilities out of our hands.

we talked about it offline, I understand why we can't use it now. 
I would reconsider doing it once/if we ever renounce our rback responsibilities.

RamLavi reviewed (approved):
/lgtm

EdDev reviewed (approved):
Thank you!
